src/main/scala/util/BundleMap.scala

```diff
@@ -37,6 +44,81 @@ class BundleMap(val fields: Seq[BundleField]) extends Record {
   require(fields.map(_.key.name).distinct.size == fields.size)
   override def cloneType: this.type = (new BundleMap(fields)).asInstanceOf[this.type]
   val elements = ListMap(fields.map { bf => bf.key.name -> chisel3.experimental.DataMirror.internal.chiselTypeClone(bf.data) } :_*)
-  def apply[T <: Data](key: BundleKey[T]) = elements(key.name).asInstanceOf[T]
+
+  def apply[T <: Data](key: BundleKey[T]): T         = elements(key.name).asInstanceOf[T]
+  def lift [T <: Data](key: BundleKey[T]): Option[T] = elements.lift(key.name).map(_.asInstanceOf[T])
+
+  def apply(key: BundleKeyBase): Data         = elements(key.name)
+  def lift (key: BundleKeyBase): Option[Data] = elements.lift(key.name)
+
+  // For every key where this.output or that.input
+  def assignL(that: BundleMap): Seq[BundleKeyBase] = {
+    val a = this.fields.filter( _.isOutput).map(_.key)
+    val b = that.fields.filter(!_.isOutput).map(_.key)
+    (a ++ b).toList.distinct
+  }
+
+  // For every key where this.input or that.output
+  def assignR(that: BundleMap): Seq[BundleKeyBase] = {
+    val a = this.fields.filter(!_.isOutput).map(_.key)
+    val b = that.fields.filter( _.isOutput).map(_.key)
+    (a ++ b).toList.distinct
+  }
 }
 
+// Implement :<= :=> and :<>
+object FixChisel3 {
```

Comment:
This would probably be better as `private[freechips]` if you are proposing something you want to get into Chisel3.

Reply:
Could you elaborate? I'm using these operators in other repositories, too. Are you proposing some form of upgrade path?

Reply:
My understanding of this is that this PR is a proposed API addition to Chisel3, but has immediate need and is too high priority to go into Chisel3 first. So, the upgrade path should look like this:

1) `object FixChisel3` is added to Rocket Chip
2) A PR adding `:<=`, `:=>`, and `:<>` is filed on Chisel3 and gets into the merge window for 3.3 (closing in ~January)
3) Rocket bumps to 3.3 at the end of January (or sooner if comfortable with master) and `object FixChisel3` is deprecated and removed.

The best situation is to limit how far `FixChisel3` gets in terms of usage to limit the deprecation/removal of (3). One way to do that is to limit this to `private[freechips]`. However, if you need this in `sifive` packages, that won't work.

Alternatively, this is not intended to ever go into Chisel3 and/or not something SiFive wants to contribute, then there's no need to limit the scope.

Reply:
I am using these operators in other packages, yes. I can't imagine removing this object in the same PR that bumps chisel would be a huge problem, would it?

Reply:
It's an API breakage. I was hoping there might be a way to avoid that. Sounds like there isn't. :( No reason to hold this up.

Reply:
I was thinking about this and I think it's actually perfectly safe to add methods that are previously defined using implicits. The only catch is the methods must be added to the base class rather than using an implicit conversion, but any new connection operators should be added in such a fashion. See this Scastie worksheet for an example: https://scastie.scala-lang.org/dGdCMA2NS3GwyIyDNzI7uQ